services/src/main/java/org/keycloak/vault/DefaultVaultTranscriber.java

```diff
@@ -0,0 +1,106 @@
+import java.nio.CharBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Default {@link VaultTranscriber} implementation that uses the configured {@link VaultProvider} to obtain raw secrets
+ * and convert them into other types. By default, the {@link VaultProvider} provides raw secrets through a {@link ByteBuffer}.
+ * This class offers methods to convert the raw secrets into other types (such as {@link VaultCharSecret} or {@link WeakReference<String>}).
+ *
+ * @see VaultRawSecret
+ * @see VaultCharSecret
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public class DefaultVaultTranscriber implements VaultTranscriber {
+
+    private static final Pattern pattern = Pattern.compile("^\\$\\{vault\\.(.+)}$");
+
+    private final VaultProvider provider;
+
+    public DefaultVaultTranscriber(final VaultProvider provider) {
+        this.provider = provider;
+    }
+
+    @Override
+    public VaultRawSecret getRawSecret(final String value) {
+        String entryId = this.getVaultEntryKey(value);
+        if (entryId != null) {
+            if (this.provider == null) {
+                throw new IllegalStateException("Required VaultProvider not found - check vault configuration");
+            }
+            // we have a valid ${vault.<KEY>} string, use the provider to retrieve the entry.
+            return this.provider.obtainSecret(entryId);
+        } else {
+            // not a vault expression - encode the value itself as a byte buffer.
+            ByteBuffer buffer = value != null ? ByteBuffer.wrap(value.getBytes(StandardCharsets.UTF_8)) : null;
+            return DefaultVaultRawSecret.forBuffer(Optional.ofNullable(buffer));
+        }
+    }
+
+    @Override
+    public VaultCharSecret getCharSecret(final String value) {
+        // obtain the raw secret and convert it into a char secret.
+        try (VaultRawSecret rawSecret = this.getRawSecret(value)) {
+            if (!rawSecret.get().isPresent()) {
+                return DefaultVaultCharSecret.forBuffer(Optional.empty());
+            }
+            ByteBuffer rawSecretBuffer = rawSecret.get().get();
+            CharBuffer charSecretBuffer = StandardCharsets.UTF_8.decode(rawSecretBuffer);
+            return DefaultVaultCharSecret.forBuffer(Optional.of(charSecretBuffer));
+        }
+    }
+
+    @Override
+    public VaultStringSecret getStringSecret(final String value) {
+        // obtain the raw secret and convert it into a string string.
+        try (VaultRawSecret rawSecret = this.getRawSecret(value)) {
+            if (!rawSecret.get().isPresent()) {
+                return DefaultVaultStringSecret.forString(Optional.empty());
+            }
+            ByteBuffer rawSecretBuffer = rawSecret.get().get();
+            return DefaultVaultStringSecret.forString(Optional.of(StandardCharsets.UTF_8.decode(rawSecretBuffer).toString()));
```

Comment:
```suggestion
            return DefaultVaultStringSecret.forString(Optional.of(new String(rawSecretBuffer, StandardCharsets.UTF_8)));
```

Reply:
Can't build a string using the ByteBuffer as a parameter, only bytes. And we can't just do rawSecretBuffer.array() because that can be null when a direct buffer is used, so the right way to obtain the internal array would be to create a byte[] with the buffer capacity and use rawSecretBuffer.get(byte[]). Not sure it is worth doing it here though.

Reply:
Ah, thanks for explanation, that makes sense